charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml

```diff
@@ -36,6 +36,110 @@ groups:
   - record: seed:hvpa_status_applied_vpa_recommendation_target:sum_by_container
     expr: sum(hvpa_status_applied_vpa_recommendation{recommendation="target"}) by (container, resource)
 
+  # Recording rules for the count of containers which have CPU usage at or above 80% of their request, aggregated by containers.
+  - record: seed:containers_over_cpu_scaling_threshold_total:count_by_container
+    expr: count(max(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / max(kube_pod_container_resource_requests_cpu_cores) by (container, namespace) >= 0.8) by (container)
+
+  # Recording rules for the count of containers which have memory usage at or above 120% of their request, aggregated by containers.
+  - record: seed:containers_over_memory_scaling_threshold_total:count_by_container
+    expr: count(max(container_memory_working_set_bytes) by (container, namespace) / max(kube_pod_container_resource_requests_memory_bytes) by (container, namespace) >= 1.2) by (container)
```

Comment:
As discussed separately, since this is supposed to provide info on HPA scaling, it should use average utilisation across namespace, and not max.
```
count(avg(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / max(kube_pod_container_resource_requests_cpu_cores) by (container, namespace) >= 0.8) by (container)
```
or
```
count(sum(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / sum(kube_pod_container_resource_requests_cpu_cores) by (container, namespace) >= 0.8) by (container)
```

Also, it might be better to filter based on relevant container, and use for eg. `container_cpu_usage_seconds_total{container="kube-apiserver"}`, so that these calculations are not done for all the control plane containers. From what I have observed, letting prometheus do lot of calculations triggers VPA update on prometheus making it unavailable for several minutes.

Also, since the thresholds (80%,120%) used in these expressions are very specific to apiserver's HPA, it makes more sense to filter based on only apiserver container.

It may also be worthwhile to add a comment somewhere that these expressions provide misleading data when apiserver is going through rolling update because of vertical scale up.

Reply:
Already made the change yesterday. But forgot to push. Done now. I have also added comments throughout (the same as what I mentioned in this PR discussion). PTAL.